cmd/metrics-v2.go

```diff
@@ -0,0 +1,1186 @@
+					Value:          float64(disk.UsedSpace),
+					VariableLabels: map[string]string{"disk": disk.DrivePath},
+				})
+
+				metrics.Metrics = append(metrics.Metrics, Metric{
+					Description:    getNodeDiskFreeBytesMD(),
+					Value:          float64(disk.AvailableSpace),
+					VariableLabels: map[string]string{"disk": disk.DrivePath},
+				})
+
+				metrics.Metrics = append(metrics.Metrics, Metric{
+					Description:    getNodeDiskTotalBytesMD(),
+					Value:          float64(disk.TotalSpace),
+					VariableLabels: map[string]string{"disk": disk.DrivePath},
+				})
+			}
+		},
+	}
+}
+func getClusterStorageMetrics() MetricsGroup {
+	return MetricsGroup{
+		Metrics: []Metric{},
+		initialize: func(ctx context.Context, metrics *MetricsGroup) {
+
+			objLayer := newObjectLayerFn()
+			// Service not initialized yet
+			if objLayer == nil {
+				return
+			}
+
+			// Fetch disk space info, ignore errors
+			storageInfo, _ := objLayer.StorageInfo(GlobalContext)
+			onlineDisks, offlineDisks := getOnlineOfflineDisksStats(storageInfo.Disks)
+			totalDisks := offlineDisks.Merge(onlineDisks)
+
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterCapacityTotalBytesMD(),
+				Value:       float64(GetTotalCapacity(ctx)),
+			})
+
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterCapacityFreeBytesMD(),
+				Value:       float64(GetTotalCapacityFree(ctx)),
+			})
+
+			s, _ := objLayer.StorageInfo(GlobalContext)
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterCapacityUsageBytesMD(),
+				Value:       GetTotalUsableCapacity(ctx, s),
+			})
+
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterCapacityUsageFreeBytesMD(),
+				Value:       GetTotalUsableCapacityFree(ctx, s),
+			})
+
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterDiskOfflineTotalMD(),
+				Value:       float64(offlineDisks.Sum()),
+			})
+
+			metrics.Metrics = append(metrics.Metrics, Metric{
+				Description: getClusterDiskOnlineTotalMD(),
+				Value:       float64(totalDisks.Sum()),
+			})
+		},
+	}
+}
+
+func (b *BucketUsageInfo) hasReplicationUsage() bool {
+	return b.ReplicationPendingSize > 0 ||
+		b.ReplicationFailedSize > 0 ||
+		b.ReplicatedSize > 0 ||
+		b.ReplicaSize > 0
+}
+
+type minioClusterCollector struct {
+	desc *prometheus.Desc
+}
+
+func newMinioClusterCollector() *minioClusterCollector {
+	return &minioClusterCollector{
+		desc: prometheus.NewDesc("minio_stats", "Statistics exposed by MinIO server", nil, nil),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+func (c *minioClusterCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (c *minioClusterCollector) Collect(ch chan<- prometheus.Metric) {
+
+	publish := func(wg *sync.WaitGroup, in <-chan Metric, out chan<- prometheus.Metric) {
```

Comment:
wait group is not a pointer it is already a pointer by just using as `var wg sync.WaitGroup`